examples/vision/token_learner.py

```diff
@@ -0,0 +1,517 @@
+references:
+
+* [Official TokenLearner code](https://github.com/google-research/scenic/blob/main/scenic/projects/token_learner/model.py)
+* [Image Classification with ViTs on keras.io](https://keras.io/examples/vision/image_classification_with_vision_transformer/)
+* [TokenLearner slides from NeurIPS 2021](https://nips.cc/media/neurips-2021/Slides/26578.pdf)
+"""
+
+"""
+## Setup
+
+We will need to install TensorFlow Addons to run this example. To install it, execute the
+following:
+
+```shell
+pip install tensorflow-addons
+```
+"""
+
+"""
+## Imports
+"""
+
+import tensorflow as tf
+
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow_addons as tfa
+
+from datetime import datetime
+import matplotlib.pyplot as plt
+import numpy as np
+
+import math
+
+"""
+## Hyperparameters
+
+Please feel free to change the hyperparameters and check your results. The best way to
+get an intuition about the architecture is to experiment with it.
+"""
+
+# DATA
+BATCH_SIZE = 256
+AUTO = tf.data.AUTOTUNE
+INPUT_SHAPE = (32, 32, 3)
+NUM_CLASSES = 10
+
+# OPTIMIZER
+LEARNING_RATE = 1e-3
+WEIGHT_DECAY = 1e-4
+
+# TRAINING
+EPOCHS = 20
+
+# AUGMENTATION
+IMAGE_SIZE = 48  # We will resize input images to this size.
+PATCH_SIZE = 6  # Size of the patches to be extracted from the input images.
+NUM_PATCHES = (IMAGE_SIZE // PATCH_SIZE) ** 2
+
+# ViT ARCHITECTURE
+LAYER_NORM_EPS = 1e-6
+PROJECTION_DIM = 128
+NUM_HEADS = 4
+NUM_LAYERS = 4
+MLP_UNITS = [
+    PROJECTION_DIM * 2,
+    PROJECTION_DIM,
+]
+
+# TOKENLEARNER
+NUM_TOKENS = 4
+
+"""
+## Load and prepare the CIFAR-10 dataset
+"""
+
+# Load the CIFAR-10 dataset.
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
+(x_train, y_train), (x_val, y_val) = (
+    (x_train[:40000], y_train[:40000]),
+    (x_train[40000:], y_train[40000:]),
+)
+print(f"Training samples: {len(x_train)}")
+print(f"Validation samples: {len(x_val)}")
+print(f"Testing samples: {len(x_test)}")
+
+# Convert to tf.data.Dataset objects.
+train_ds = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+train_ds = train_ds.shuffle(BATCH_SIZE * 100).batch(BATCH_SIZE).prefetch(AUTO)
+
+val_ds = tf.data.Dataset.from_tensor_slices((x_val, y_val))
+val_ds = val_ds.batch(BATCH_SIZE).prefetch(AUTO)
+
+test_ds = tf.data.Dataset.from_tensor_slices((x_test, y_test))
+test_ds = test_ds.batch(BATCH_SIZE).prefetch(AUTO)
+
+"""
+## Data augmentation
+
+The augmentation pipeline consists of:
+
+- Rescaling
+- Resizing
+- Random cropping (fixed-sized or random sized)
+- Random horizontal flipping
+"""
+
+data_augmentation = keras.Sequential(
+    [
+        layers.Rescaling(1 / 255.0),
+        layers.Resizing(INPUT_SHAPE[0] + 20, INPUT_SHAPE[0] + 20),
+        layers.RandomCrop(IMAGE_SIZE, IMAGE_SIZE),
+        layers.RandomFlip("horizontal"),
+    ],
+    name="data_augmentation",
+)
+
+"""
+Note that these layers have their inference behaviour pre-defined. Refer
```

Comment:
The meaning of the sentence "these layers have their inference behaviour pre-defined" is unclear. Do this refer to the fact that augmentation layers (above, that's `RandomCrop` and `RandomFlip`) are pass-through when `training=False`?

Reply:
That's correct. However, RandomCrop's inference behavior is a little different. From the [official docs](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_crop/):

> At inference time, the images will be first rescaled to preserve the shorter side, and center cropped.